### PR TITLE
[improve][broker] Optimize topic production rate stats.

### PR DESCRIPTION
**Pull request by hanmz**
### Motivation

In a production environment, we often encounter a scenario: send a message to create a producer, and then close it after use. At this time, we will see that the `msgRateIn` and `msgThroughputIn` in topics stats are always 0. This is because the `msgRateIn` and `msgThroughputIn` of the topic are calculated from the online producer.
(Here I only optimized the production rate stats, and the consumption rate stats also have similar problems.)

### Modifications
1. Add `rateIn` statistics in topic to avoid relying on producers being online.
2. Add an optional configuration to be compatible with the previous logic.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java [hunk 1]
```diff
@@ -3449,6 +3449,13 @@ The max allowed delay for delayed delivery (in milliseconds). If the broker rece
     )
     private Set<String> additionalServlets = new TreeSet<>();
 
+    @FieldContext(
+            category = CATEGORY_SERVER,
+            dynamic = true,
+            doc = "Enable or disable replace producer stats with topic stats when calculating topic production rate"
+    )
+    private boolean enableReplaceProducerStatsWithTopicStats = false;
```
Comment (heesung-sn) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java:3457]: `precomputeProducerStatsInTopicStats` might be the better name here.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 2]
```diff
@@ -2111,6 +2111,10 @@ public void updateRates(NamespaceStats nsStats, NamespaceBundleStats bundleStats
         bundleStats.producerCount += producers.size();
         topicStatsStream.startObject(topic);
 
+        if (brokerService.pulsar().getConfig().isEnableReplaceProducerStatsWithTopicStats()) {
+            rateIn.calculateRate();
```
Comment (heesung-sn) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2115]: It seems like we don't skip/update topicStatsHelper.aggMsgRateIn and aggMsgThroughputIn here.

**Review by heesung-sn — COMMENTED** [review]
Plz add unit tests to cover this logic.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java [hunk 3]
```diff
@@ -885,6 +887,13 @@ public void recordAddLatency(long latency, TimeUnit unit) {
         PUBLISH_LATENCY.observe(latency, unit);
     }
 
+    @Override
+    public void recordRateIn(long events, long totalValue) {
+        if (brokerService.pulsar().getConfig().isEnableReplaceProducerStatsWithTopicStats()) {
```
[the remark this answers is not in the document]
Comment (hanmz) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java:892]: alright

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 4]
```diff
@@ -2391,6 +2395,12 @@ public CompletableFuture<? extends TopicStatsImpl> asyncGetStats(GetStatsOptions
             }
         });
 
+        // Replace producer stats with topic-level stats
```
[the remark this answers is not in the document]
Comment (hanmz) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2398]: all right

**Comment by hanmz** [comment]
> Plz add unit tests to cover this logic.


I have added unit tests, PLAT~

**Comment by merlimat** [comment]
@hanmz This is a long-standing issue and the plan was to address as part of the overall changes of [PIP-264](https://github.com/apache/pulsar/blob/master/pip/pip-264.md). I wouldn't introduce a fix for just 1 metric with one dedicated config option when these metrics are going away soon. @dragosvictor has already started converting the metrics.

**Comment by hanmz** [comment]
> @hanmz This is a long-standing issue and the plan was to address as part of the overall changes of [PIP-264](https://github.com/apache/pulsar/blob/master/pip/pip-264.md). I wouldn't introduce a fix for just 1 metric with one dedicated config option when these metrics are going away soon. @dragosvictor has already started converting the metrics.

If the community will solve this problem in the near future, this PR will not be introduced.

**Pull request closed without merging**